### PR TITLE
CFE-3197: New version convenience functions

### PR DESCRIPTION
**Pull request by karlhto**
Ticket: CFE-3197
Changelog: None

### CODE REVIEW

File: libpromises/evalfunction.c [hunk 1]
```diff
@@ -2067,6 +2067,103 @@ static FnCallResult FnCallClassify(EvalContext *ctx,
     return FnReturnContext(is_defined);
 }
 
+/*********************************************************************/
+
+// Calculates major.minor.build version "value"
```
Comment (nickanderson) [inline libpromises/evalfunction.c]: s/build/patch/

**Comment by karlhto** [comment]
Still lacks acceptance tests.

File: libpromises/evalfunction.c [hunk 2]
```diff
@@ -2066,6 +2066,144 @@ static FnCallResult FnCallClassify(EvalContext *ctx,
...
+    assert(ver_string != NULL);
+
+    int mod = 10000;
+    char *endptr = NULL;
+    char *str_ptr = ver_string;
+    long res = 0;
+    for (int i = 0; i < 3; i++)
+    {
+        long ver = strtoul(str_ptr, &endptr, 10);
+        res += ver * mod;
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: I don't think you should do it this way. Split the number in 3 (major, minor, patch). Then just compare major first, then minor, then patch.

Please take a look at `macro_if_version` in `cf3lex.l`, and try to align with how that works (maybe even make a shared function from the logic there).

**Comment by karlhto** [comment]
Thanks for the nice version comparison functions, @olehermanse ! : )

**Comment by karlhto** [comment]
@cf-bottom Jenkins please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4551/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4551/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4551/

File: libpromises/evalfunction.c [hunk 3]
```diff
@@ -9429,6 +9553,12 @@ const FnCallType CF_FNCALL_TYPES[] =
                   FNCALL_OPTION_NONE, FNCALL_CATEGORY_DATA, SYNTAX_STATUS_NORMAL),
     FnCallTypeNew("concat", CF_DATA_TYPE_STRING, CONCAT_ARGS, &FnCallConcat, "Concatenate all arguments into string",
                   FNCALL_OPTION_VARARG, FNCALL_CATEGORY_DATA, SYNTAX_STATUS_NORMAL),
+    FnCallTypeNew("cf_version_minimum", CF_DATA_TYPE_CONTEXT, CFVERSION_ARGS, &FnCallVersionMinimum, "True if local CFEngine version is newer than or equal to input",
+                  FNCALL_OPTION_NONE, FNCALL_CATEGORY_UTILS, SYNTAX_STATUS_NORMAL),
+    FnCallTypeNew("cf_version_maximum", CF_DATA_TYPE_CONTEXT, CFVERSION_ARGS, &FnCallVersionMaximum, "True if local CFEngine version is older than or equal to input",
+                  FNCALL_OPTION_NONE, FNCALL_CATEGORY_UTILS, SYNTAX_STATUS_NORMAL),
+    FnCallTypeNew("cf_version_between", CF_DATA_TYPE_CONTEXT, CFVERSIONBETWEEN_ARGS, &FnCallVersionBetween, "True if local CFEngine version is between two input versions",
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: could add `between (inclusive)` to the explanation.

**Comment by karlhto** [comment]
@cf-bottom Jenkins please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4595/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4595/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4595/

File: libpromises/evalfunction.c [hunk 4]
```diff
@@ -2066,6 +2067,157 @@ static FnCallResult FnCallClassify(EvalContext *ctx,
...
+        const Rlist *args)
+{
+    const VersionComparison comparison = GenericVersionCheck(fp, args);
+    if (comparison == VERSION_ERROR)
+    {
+        return FnFailure();
+    }
+
+    return FnReturnContext((comparison == VERSION_GREATER ||
+                            comparison == VERSION_EQUAL));
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: Why the double parens? You could group the 2 comparisons, if you want, but having double parens at the outer level just looks weird.

File: libpromises/evalfunction.c [hunk 5]
```diff
@@ -2066,6 +2067,157 @@ static FnCallResult FnCallClassify(EvalContext *ctx,
...
+{
+    assert(fp != NULL);
+    assert(fp->name != NULL);
+    if (args == NULL)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Policy fuction %s requires version to compare against",
+            fp->name);
+        return VERSION_ERROR;
+    }
```
Comment (olehermanse) [inline libpromises/evalfunction.c]: Very nice, this helper function makes the code very easy to read and follow. 👏 👏 👏

File: tests/acceptance/01_vars/02_functions/cf_version.cf [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    version => "1.0";
+}
+
+###########################################################
+
+bundle agent test
+{
+  vars:
+    "after_newer" string => "yes",
+      if => cf_version_after("1.0.0");
```
Comment (olehermanse) [inline tests/acceptance/01_vars/02_functions/cf_version.cf]: Could add some test cases where you don't specify patch version.

`if => cf_version_after("3.12")`

`if => cf_version_at("3")`

Comment (karlhto) [reply tests/acceptance/01_vars/02_functions/cf_version.cf]: Regarding the second one: would be nice to make sure that the tests work between major functions though, right?

**Comment by karlhto** [comment]
@cf-bottom Jenkins

Followed up with requested changes, want Jenkins again because of acceptance test changes.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4623/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4623/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4623/

File: tests/acceptance/01_vars/02_functions/cf_version.cf (hunk 6 above)
Comment (olehermanse) [reply tests/acceptance/01_vars/02_functions/cf_version.cf]: Could be, but not a big deal, imo.

**Pull request merged**